presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/OptimizeMixedDistinctAggregations.java

```diff
@@ -0,0 +1,507 @@
+    )
+    {
+        ImmutableMap.Builder<Symbol, Aggregation> assignments = ImmutableMap.builder();
+
+        for (Map.Entry<Symbol, FunctionCall> entry : aggregateInfo.getAggregations().entrySet()) {
+            FunctionCall functionCall = entry.getValue();
+            if (!functionCall.isDistinct()) {
+                Symbol newSymbol = symbolAllocator.newSymbol(entry.getKey().toSymbolReference(), symbolAllocator.getTypes().get(entry.getKey()));
+                aggregationOutputSymbolsMapBuilder.put(newSymbol, entry.getKey());
+                if (!duplicatedDistinctSymbol.equals(distinctSymbol)) {
+                    // Handling for cases when mask symbol appears in non distinct aggregations too
+                    // Now the aggregation should happen over the duplicate symbol added before
+                    if (functionCall.getArguments().contains(distinctSymbol.toSymbolReference())) {
+                        ImmutableList.Builder arguments = ImmutableList.builder();
+                        for (Expression argument : functionCall.getArguments()) {
+                            if (distinctSymbol.toSymbolReference().equals(argument)) {
+                                arguments.add(duplicatedDistinctSymbol.toSymbolReference());
+                            }
+                            else {
+                                arguments.add(argument);
+                            }
+                        }
+                        functionCall = new FunctionCall(functionCall.getName(), functionCall.getWindow(), false, arguments.build());
+                    }
+                }
+                assignments.put(newSymbol, new Aggregation(
+                        functionCall,
+                        aggregateInfo.getFunctions().get(entry.getKey()),
+                        Optional.empty()));
+            }
+        }
+        return new AggregationNode(
+                idAllocator.getNextId(),
+                groupIdNode,
+                assignments.build(),
+                ImmutableList.of(ImmutableList.copyOf(groupByKeys)),
+                SINGLE,
+                originalNode.getHashSymbol(),
+                Optional.empty());
+    }
+
+    private Signature getFunctionSignature(QualifiedName functionName, Symbol argument, SymbolAllocator symbolAllocator)
+    {
+        return functionRegistry
+                .resolveFunction(
+                        functionName,
+                        ImmutableList.of(new TypeSignatureProvider(symbolAllocator.getTypes().get(argument).getTypeSignature())));
+    }
+
+    /*
+     * This Project is useful for cases when we aggregate on distinct and non-distinct values of same symbol, eg:
+     *  select a, sum(b), count(c), sum(distinct c) group by a
+     * Without this Project, we would count additional values for count(c)
+     *
+     * This method also populates maps of old to new symbols. For each key of outputNonDistinctAggregateSymbols,
```

Comment:
Same here.... passing in arguments to be modified requires anyone who touches the code to be aware of all the state that's going on everywhere, so it makes it harder to understand and really easy for other people to come in and mess up.